boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,293 @@
+
+    """
+
+    def __init__(
+        self,
+        current_heading: Scalar,
+        desired_heading: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        max_angle_range: tuple,
+    ):
+        """Initializes the class attributes. Note that this class cannot be directly instantiated.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees.
+            `desired_heading` (Scalar): Target boating heading direction in degrees.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `prev_control` (list): Timeseries of control angles computed over time.
+            `time_series` (list): Timeseries of running time values related to control angles.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range
+        """
+        self.current_heading = current_heading
+        self.desired_heading = desired_heading
+        self.current_control_ang = current_control_ang
+        self.time_step = time_step
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range
+        self.prev_control: List[Scalar] = list()
+        self.time_series: List[Scalar] = list()
+
+        self.time_series.append[0]  # append initial starting time 0
+        self.prev_control.append[self.current_control_ang]  # append starting control angle
+
+    def append_heading(self, past_heading: Scalar) -> None:
+        """Appends the previous heading to heading list and updates the current
+        time taken to iterate to the current rudder position
+
+        Args:
+            `past_heading` (Scalar): The previous heading to be added
+
+        """
+        self.prev_control.append(past_heading)
+        self.time_series.append(self.time_series[-1] + self.time_step)
+
+    def reset_desired_heading(self, target_angle: Scalar) -> None:
+        """Resets a new desired heading angle and clears all previous
+        headings from current_control_ang list and running time for time_series
+
+        Args:
+            `target_angle` (Scalar): New desired heading
+
+        """
+        self.desired_heading = target_angle
+        self.prev_control.clear()
+        self.time_series.clear()
+
+    def get_heading(self) -> None:
+        """Returns the most recent heading and current running time"""
+        return (self.prev_control[-1], self.time_series[-1])
+
+    @abstractmethod
+    def update_state(self, speed: Scalar) -> None:
+        """Updates the controller position based on target heading
+
+        Args:
+            `speed` (Scalar): Iterative speed in degrees / second
+
+        """
```

Comment:
Sounds good I can refactor speed in the init. I think I was considering it being individual in the function because I was considering that the speed could change in each iteration loop. As in going from 5 - 10  degrees, there might not be a constant speed (might want to start faster and end slower or something like that) without changing the internal speed. Am I overcomplicating it a bit?